File: homework1/jesus_romero/q2_ReverseVowels.py
"""
Time, Space complexities: O(n), O(n)

Q2: Reverse Vowels
Given a string, reverse the order of the vowels in the string
"""

def reverseVowels(s):
    #1. Create needed variables, a list of all characters in the string, 
    # two pointers, and a set with all vowels 

    s_list = list(s)
    low = 0
    high = len(s) - 1
    vowels = set("aeiouAEIOU")

    #2. Iterate while the two pointers don't meet
    while low < high:
        #3. If low is not a vowel, move right
        if s_list[low] not in vowels:
            low += 1
        #4. if high is not a vowel, move left
        elif s_list[high] not in vowels:
            high -= 1
        #5. If both are vowels, swap them and move both respectively 
        else:
            temp = s_list[low]
            s_list[low] = s_list[high]
            s_list[high] = temp

            low += 1
            high -= 1

    #6. Join the resulting list into a string
    return "".join(s_list)

def test_reverseVowels():
    input_string = "Uber Career Prep"
    expected = "eber Ceraer PrUp"
    result = reverseVowels(input_string)
    assert result == expected, f"Expected {expected}, got {result}"

    input_string = "xyz"
    expected = "xyz"
    result = reverseVowels(input_string)
    assert result == expected, f"Expected {expected}, got {result}"

    input_string = "flamingo"
    expected = "flominga"
    result = reverseVowels(input_string)
    assert result == expected, f"Expected {expected}, got {result}"


if __name__ == "__main__":
    test_reverseVowels()

File: homework1/tom_nguyen/q2_ReverseVowels.cpp
#include <iostream>
#include <string>
#include <unordered_set>
using namespace std;

//uses two pointers to reverse the order of vowels in a string in O(n)
string reverseVowels(string s) {
    unordered_set<char> vowels = {'a', 'e', 'i', 'o', 'u', 'A', 'E', 'I', 'O', 'U'};
    int left = 0;
    int right = s.size() - 1;
    while (left < right) { //move pointers to vowels index
        if (!vowels.contains(s[left])) {
            ++left;
            continue;
        }
        if (!vowels.contains(s[right])) {
            --right;
            continue;
        }
        swap(s[left], s[right]);
        ++left;
        --right;
    }
    return s;
}

int main() {
    cout << reverseVowels("Uber Career Prep") << endl; // eber Ceraer PrUp
    cout << reverseVowels("xyz") << endl; // xyz
    cout << reverseVowels("flamingo") << endl; // flominga

    return 0;
}

//time spent: 5 minutes
